common/web/types/src/ldml-keyboard/pattern-parser.ts

```diff
@@ -56,6 +82,35 @@ export class MarkerParser {
     }
     return matchArray(str, this.REFERENCE);
   }
+
+  /** @returns string for marker #n */
+  public static markerOutput(n: number): string {
+    if (n < MarkerParser.MIN_MARKER_INDEX || n > MarkerParser.ANY_MARKER_INDEX) {
+      throw RangeError(`Internal Error: marker index out of range ${n}`);
+    }
+    return this.SENTINEL + String.fromCharCode(n);
+  }
+
+  /** @returns all marker strings as sentinel values */
+  public static toSentinelString(s: string, markers?: OrderedStringList) : string {
+    if (!s) return s;
+    return s.replaceAll(this.REFERENCE, (sub, arg) => {
+      if (arg === MarkerParser.ANY_MARKER_ID) {
+        return MarkerParser.SENTINEL_ALL_MARKERS;
+      }
+      if (!markers) {
```

Comment:
Given this, why is `markers` an optional parameter?

Reply:
May be no markers in that file.